buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/license/LicenseResolver.groovy

```diff
@@ -0,0 +1,91 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.gradle.plugins.license
+
+import groovy.transform.CompileStatic
+import org.gradle.api.Project
+import org.gradle.api.artifacts.Configuration
+import org.gradle.api.artifacts.Dependency
+import org.gradle.api.artifacts.ProjectDependency
+import org.gradle.api.publish.Publication
+import org.gradle.api.publish.PublishingExtension
+import org.gradle.api.publish.maven.MavenPublication
+import org.gradle.api.publish.maven.internal.publication.DefaultMavenPom
+
+@CompileStatic
+class LicenseResolver {
+
+    /** Resolve all transitive licenses. */
+    Set<ProjectLicense> resolve(Project project, Configuration configuration) {
+        return determineExternalLicenses(project, configuration) + determineProjectLicenses(configuration)
+
+    }
+
+    private static Set<ProjectLicense> determineProjectLicenses(Configuration configuration) {
+        configuration.allDependencies.findAll {
+            it instanceof ProjectDependency
+        }.collectMany {
+            Project depProject = ((ProjectDependency) it).dependencyProject
+            Set<ProjectLicense> transitive = determineProjectLicenses(depProject.configurations.getByName(configuration.name))
+            return [determineLicense(depProject)] + transitive
+        }.findAll { it != null } as Set
+    }
+
+    private static Set<ProjectLicense> determineExternalLicenses(Project project, Configuration configuration) {
+        def conf = project.configurations.create('internalAllExternalArtifacts')
+
+        configuration.allDependencies.each {
```

Comment:
Four our future selves, can you leave a comment on why these tactics are necessary?